### PR TITLE
Metrics via dagger

### PR DESCRIPTION
**Pull request by jflo**
Introduces Dagger to provide dependencies.  

- Migrates an existing MetricsSystem module up into the metrics gradle module.
- Makes BesuCommand accessible from Dagger components, to enable configuration reuse.
- Uses CachedMerkleTrieLoader as an example of how to get and use a MetricSystem.

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/Besu.java
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/Besu.java:19]: Presume this is generated by Dagger?

File: besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java:39]: BesuComponent seems quite generic to me. Could/should we be more precise with the name? Or does it refer to the one and only component for the `besu` gradle module? 
Will this end up being _the_ application context, or one of many (e.g. one per module)?

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1338]: What's the difference between BesuComponent and DaggerBesuComponent?

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:479]: comment should be removed?

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:513]: In what scenario would besuComponent be null here?

File: besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java
Comment (siladu) [inline besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java:171]: mix of 'Context' and 'Component' being used in the var names

File: ethereum/referencetests/src/reference-test/external-resources
Comment (siladu) [inline ethereum/referencetests/src/reference-test/external-resources:1]: see my message about fixing this on #besu-contributors

File: gradle.properties
Comment (siladu) [inline gradle.properties:5]: keep cranking it!! 😆

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/MetricsConfigurationModule.java
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/MetricsConfigurationModule.java:28]: Intrigued to know why you were able to delete this

**Review by siladu — COMMENTED** [review]
Looks pretty good to me, but I haven't fully got my head around Dagger yet so this is a first pass. Sorry for all the questions :)

Love the fact this is targetting BesuCommand...I believe that's the most coupled class?

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1338]: BesuComponent is ours, the Dagger_ prefix indicates it is the implementation that Dagger generates for us during precompile. At some point, once we've "daggerized all the things" we won't need to use these because Dagger will more fully compose the object graph.

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:513]: Good catch. This is leftover from a prior revision where we were passing in the BesuComponent after construction time.

File: besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java
Comment (jflo) [reply besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java:171]: This was intentional per the description of Component above, but I am happy to change if this is misleading.  Eventually we will introduce subcomponents, which I would name something like P2PComponent, but then refer to the top-level component as a single application context.

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/MetricsConfigurationModule.java
Comment (jflo) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/MetricsConfigurationModule.java:28]: It moved packages, now the module is in the same package that MetricsConfiguration is in.

File: besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java:39]: Any reason not to call the whole class ApplicationContext? Or is that to Spring-y 😛

File: besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java
Comment (siladu) [reply besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java:171]: I guess I'd favour whatever is idiomatic for Dagger. My objection here is about having two variable names for the same thing, would prefer if we stick with one or the other everywhere.

Comment (siladu) [reply besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java:171]: I'd be fine with any of mockContext, mockBesuContext, mockApplicationContext, mockBesuComponent, as long as we use the same in the prod code.

**Comment by siladu** [comment]
@jflo I attempted a second pass after reading up on Dagger, but I think the recent merge was dodgy and has polluted the PR files so it's hard to review now.

Are you back to working on it or is it still ready for review?

File: besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java:39]: Yes, trying to adopt Daggers convention

**Review by siladu — COMMENTED** [review]
Forgot to submit this comment yesterday

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/cache/CachedMerkleTrieLoaderModule.java
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/cache/CachedMerkleTrieLoaderModule.java:1]: does this need to be in this package for visibility reasons?  could we put all of the modules in the besu project so we can find them?

**Review by garyschulte — APPROVED** [review]
Love it.  as mentioned in comment, I wish we had a way to collect all of the module definitions though.  It would be wise to get a handful of approvals before merging since this is a 'whole repo' kind of change, even if incremental

File: besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java:37]: Would love to get more maintainers eyes on this PR, particularly regarding this question, before this is merged and a precedent for how to design modules is set :)

File: besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java
Comment (macfarla) [reply besu/src/test/java/org/hyperledger/besu/cli/CommandTestAbstract.java:172]: mockBesuComponent makes sense to me

**Review by macfarla — COMMENTED** [review]
still getting my head around dagger also. This change doesn't get us out of passing the metricsSystem around eg to transactionPool - it just makes setup easier?

**Review by gfukushima — COMMENTED** [review]
Also still getting my head around dagger. Could even try implementing another module. Do you have any suggestions of what would be a good one to get started with to familiarise with dagger?

File: besu/src/main/java/org/hyperledger/besu/Besu.java
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/Besu.java:19]: Yes, see explanation in later comment.

File: besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/components/BesuComponent.java:37]: Devs should DEFINITELY not assume this establishes a precedent of any kind. The right place to do this could be a judgement call. If something is worth being Daggerized within its submodule (evmtool is an example) then the dagger stuff should live right next to its implementation. In other cases, there could be a module that is only ever useful in a BesuContext (like configuration related things).

**Comment by jflo** [comment]
> still getting my head around dagger also. This change doesn't get us out of passing the metricsSystem around eg to transactionPool - it just makes setup easier?

Correct. Once transactionPool is itself constructed by Dagger, then we no longer need to pass it or the BesuContext around ourselves.

**Comment by jflo** [comment]
> Do you have any suggestions of what would be a good one to get started with to familiarise with dagger?

The ideal candidate would be something used rarely, but deeeeeply within the overall object graph.

**Pull request merged**
